frontend/pages/admin/IntegrationsPage/cards/Calendars/Calendars.tsx

```diff
@@ -0,0 +1,219 @@
+import React, { useState, useContext } from "react";
+import { useQuery } from "react-query";
+
+import { IConfig } from "interfaces/config";
+import { NotificationContext } from "context/notification";
+import { AppContext } from "context/app";
+import configAPI from "services/entities/config";
+
+// @ts-ignore
+import InputField from "components/forms/fields/InputField";
+import Button from "components/buttons/Button";
+import SectionHeader from "components/SectionHeader";
+import CustomLink from "components/CustomLink";
+import Spinner from "components/Spinner";
+import DataError from "components/DataError";
+import PremiumFeatureMessage from "components/PremiumFeatureMessage/PremiumFeatureMessage";
+
+const LEARN_MORE_CALENDARS =
+  "https://www.fleetdm.com/learn-more-about/google-workspace-service-accounts";
+
+interface IFormField {
+  name: string;
+  value: string | boolean | number;
+}
+
+interface ICalendarsFormErrors {
+  email?: string | null;
+  domain?: string | null;
+  privateKey?: string | null;
+}
+
+const baseClass = "calendars-form";
+
+const Calendars = (): JSX.Element => {
+  const { renderFlash } = useContext(NotificationContext);
+  const { isPremiumTier } = useContext(AppContext);
+
+  const [formData, setFormData] = useState({
+    email: "",
+    domain: "",
+    privateKey: "",
+  });
+  const [isUpdatingSettings, setIsUpdatingSettings] = useState(false);
+  const [formErrors, setFormErrors] = useState<ICalendarsFormErrors>({});
+
+  const {
+    data: appConfig,
+    isLoading: isLoadingAppConfig,
+    refetch: refetchConfig,
+    error: errorAppConfig,
+  } = useQuery<IConfig, Error, IConfig>(["config"], () => configAPI.loadAll(), {
+    select: (data: IConfig) => data,
+    onSuccess: (data) => {
+      setFormData({
+        email: data.integrations.google_calendar[0].email,
+        domain: data.integrations.google_calendar[0].domain,
+        privateKey: data.integrations.google_calendar[0].private_key,
+      });
+    },
+  });
+
+  const { email, domain, privateKey } = formData;
+
+  const handleInputChange = ({ name, value }: IFormField) => {
+    setFormData({ ...formData, [name]: value });
+    setFormErrors({});
```

Comment:
I like the validation on blur, but it means the user could see an enabled save button and click it even while a field has an invalid value, since the blur event won't happen until they've clicked the save button. Other newer forms like `TeamSettings` are validating here, within the `handleInputChange`, which I think works well. Can we do that here too?

Reply:
Thanks for the example, I'll take a look at implementation today!